### PR TITLE
Initial implementation Chip component

### LINKED ISSUE #128: Implement Chip component
Apparently underway in upstream: https://github.com/google/material-design-lite/tree/mdl-1.x/src/chip.

### PR DESCRIPTION
**Pull request by vipentti**
Initial implementation of the `Chip` component new in version 1.2.0 to resolve #128 

Demo could use some beefing up perhaps.

### CODE REVIEW

File: demo/Demo/Chips.elm [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+
+        , Chip.button []
+            [ Icon.view "cancel" [] ]
+        ]
+    , """
+      Chip.chip
+        [ Chip.deletable ]
+        [ Chip.text []
+            [ text "Deletable Chip" ]
+
```
Comment (debois) [inline demo/Demo/Chips.elm]: Unnecessary linebreak? (Also above.)

File: demo/Demo/Chips.elm [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        , Chip.button []
+            [ Icon.view "cancel" [] ]
+        ]
+       """
+    )
+  , ( Chip.chip' Html.button
```
Comment (debois) [inline demo/Demo/Chips.elm]: Neat.

File: demo/Demo/Chips.elm [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        [ Options.type' "button" ]
+        [ Chip.text []
+            [ text "Button Chip" ]
+        ]
+       """
+    )
+  ,  ( Chip.chip
+          [ Chip.contact ]
+          [ Chip.contactItem
+              [ Options.cs "mdl-color--teal"
```
Comment (debois) [inline demo/Demo/Chips.elm]: Would this work?

```elm
[ Color.background Color.primary
, Color.text Color.white 
] 
```

File: demo/Demo/Chips.elm [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+              ]
+              [ text "A" ]
+          , Chip.text []
+              [ text "Contact Chip" ]
+          ]
+     , """
+        Chip.chip
+          [ Chip.contact ]
+          [ Chip.contactItem
+              [ Options.cs "mdl-color--teal"
```
Comment (debois) [inline demo/Demo/Chips.elm]: See above.

File: demo/Demo/Chips.elm [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+              [ text "Contact Chip" ]
+          ]
+        """
+     )
+  , ( Chip.chip
+        [ Chip.contact
+        , Chip.deletable
+        ]
+        [ Chip.contactItem
+            [ Options.cs "mdl-color--teal"
```
Comment (debois) [inline demo/Demo/Chips.elm]: See above.

File: demo/Demo/Chips.elm [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+            [ Options.attribute <| Html.href "#chips" ]
+            [ Icon.view "cancel" [] ]
+        ]
+    , """
+      Chip.chip
+        [ Chip.contact
+        , Chip.deletable
+        ]
+        [ Chip.contactItem
+            [ Options.cs "mdl-color--teal"
```
Comment (debois) [inline demo/Demo/Chips.elm]: See above.

File: src/Material/Chip.elm [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+{-| Create a chip contained in the given element
+-}
+chip' : HtmlElement msg -> List (Options.Property c msg) -> List (Html msg) -> Html msg
+chip' element props content =
+  Options.styled element
+    (cs "mdl-chip" :: props)
+    content
+
+
+{-| Creates a chip using `Html.span`
```
Comment (debois) [inline src/Material/Chip.elm]: Neat.

**Comment by debois** [comment]
I like this. The API looks very nice when used in the demo.

Could you say a little bit about why Chips are implemented as element transformers and why we need to expose `chips'`?

It seems Chips could have been implemented as just a bunch of `Property c m`, which users would then have to apply correctly. (Ok, maybe that's why?)

**Comment by vipentti** [comment]
@debois I initially had them as properties but I felt the boilerplate to create a chip was too much, especially because applying them correctly would require knowledge about the structure. 

e.g. 
```elm
Options.styled Html.span
  [ Chip.chip
  , Chip.deletable
  ]
  [ Options.styled Html.span
      [ Chip.text ]
      [ text "Deletable Chip" ]
  , Options.styled' Html.button
    [ Chip.action ]
    [ Html.type' "button" ]
    [ Icon.view "cancel"
        []
    ]
  ]
```

vs 

```elm
Chip.chip
  [ Chip.deletable ]
  [ Chip.text []
      [ text "Deletable Chip" ]
  
  , Chip.button []
      [ Icon.view "cancel" [] ]
  ]
```

Also the reason I expose `chip'` and `text'` and other variants is because `Chip` is typically a `span` or `div` but it could also be a `button` and such.

**Comment by OvermindDL1** [comment]
Why not have Chip.chip take a node as the first element?  That way we could do:
```elm
Chip.chip Html.div
  [ Chip.deletable ]
  [ Chip.text [] {- I'd keep Chip.text for ease of repeatability -}
      [ text "Deletable Chip" ]
  , Chip.action Html.Button [] {- Or maybe Material.Button, or Html.href or whatever? -}
      [ Icon.view "cancel" [] ]
  ]
```

**Comment by vipentti** [comment]
@OvermindDL1 That's what `Chip.chip'` allows, we could always swap those around if needed

**Comment by OvermindDL1** [comment]
Ah, I missed that (was concentrating on other code >.>).  Looks good in that case although personally I'd rename `chip'` to `chip` and rename `chip` to `chip` to `chipSpan` or so.  I always prefer explicit over implicit in naming.  :-)

**Comment by debois** [comment]
Couple of questions: 

1. Can we get rid of `chip.deletable`? From the [upstream implementation](https://github.com/google/material-design-lite/blob/mdl-1.x/src/chip/_chip.scss#L81-L83) what it does is add padding to accommodate an icon at the right end of the chip; can we somehow roll that into the `...action` bits?

1a. Otherwise, we should probably give it a different name, since it's necessary whenever your chip has an icon, yes?

2. According to [the spec](https://material.google.com/components/chips.html#chips-usage), clicking a chip should open "a full detailed view (either in a card or full screen) or a menu of options". We can talk about `menu` when @aforemny finalises #165, but for now, is there a way for me to add an action to clicking the entire chip? If there is, can we show that in the demo?

3. I'm thinking it maybe isn't necessary to let user's customise elements used in actions (typically delete). We just need to make sure he has the option of either (a) dispatching a message, or (b) going to some URL. Is this possible?

4. I'm unsure why it's necessary to let user's decide what the top-level element of a chip should be. I can see `button` and `span` renders differently, but beyond those two options, are there other things he may want?

I was thinking about releasing a patch release shortly, but hold off Chips until we get the API _just_ right?

**Comment by vipentti** [comment]
I think we can hold off `Chips` until we finalize the API

**Comment by vipentti** [comment]
Regarding 3. I think this is where we can decide how much we want to simply follow `MDL` and how much we want to be an UI implementation that uses MDL.  

MDL says: 
> Create a container element for the chip; typically `<span>` and `<div>`
> are used, but any container element should work equally well. If you need
> interactivity, use a `<button>`, or add the tabindex attribute to your container.

I will work on adding an interactive example(s) with deletable actions and such.

**Comment by vipentti** [comment]
Did a refactor of the Chip API. It does still look somewhat the same to the user but internally there are more changes.

1. Added opaque `Content` type to `Chip` to only limit certain types of content that can be added (more strict than `MDL` but closer to the spec).

3. `deletable` is no longer a property or an action that can be provided by the user.
To make a chip "deletable" simply add one of the following to the Chip: `Chip.deleteIcon`, `Chip.deleteClick`, `Chip.deleteAction`. Adding one of these will automatically add the `mdl-chip--deletable` class as well as the icon. Icon can be changed by `Chip.deleteIcon`. By default `Icon.view "cancel" []` is used.

4. `contact` is no longer a property. It now is a function that creates a contact-element, and if such an element is present then the `mdl-chip--contact` class is added to the chip automatically. 

5. The naming of the Chip functions are still in the open. I'm not sure if the `chip` prefix is necessary for the `Chip.chipSpan` and `Chip.chipButton` types. The could probably just be named `Chip.span` and `Chip.button` (also to discourage someone just exposing everything).

6. Chips also support custom event handlers with `Chip.on`. There is really no particular reason for it.

Some examples: 
```elm
Chip.chipSpan
  [ Chip.deleteIcon "cancel" ]
  [ Chip.content []
      [ text "Deletable Chip" ]
  ]

Chip.chipSpan
  [ Chip.deleteLink "#chips"
  ]
  [ Chip.contact Html.span
      [ Color.background Color.primary
      , Color.text Color.white
      ]
      [ text "A" ]
  , Chip.content []
      [ text "Deletable Contact Chip" ]
  ]

Chip.chipButton
  [ Chip.onClick (ChipClick index)
  , Chip.deleteClick (RemoveChip index)
  ]
  [ Chip.content []
      [ text ("Amazing Chip " ++ toString index ) ]
  ]
```


As always looking for feedback and suggestions.

**Comment by OvermindDL1** [comment]
+1 vote for removing the `chip` prefix to become `Chip.span` and `Chip.button` and such as well.


Is there any chance for this Chip container thing that is listed in the animation at https://material.google.com/components/chips.html#chips-behavior so we could, for example, make adding users, tags, etc... like how it is in GMail, type a name in a textbox, auto-completes out and it resolves into a Chip?


As per https://material.google.com/components/chips.html#chips-contact-chips what is the way to pop-up the extra details box for a user (such as in the linked example of being able to choose from a few different email addresses via the Material defined Normal, Focused, Pressed, and Activated states (are able to active multiple in some implementations I've seen as well).

**Comment by vipentti** [comment]
The current upstream MDL does not implement the full Contact chip specification; missing Open contact chip. It might be possible to implement it with Lists  and some absolutely positioned container. Something like how tooltip works. 

As for the resolving to Chips, it should be doable just having a container with the chips and a textfield.

**Comment by OvermindDL1** [comment]
> The current upstream MDL does not implement the full Contact chip specification; missing Open contact chip. It might be possible to implement it with Lists and some absolutely positioned container. Something like how tooltip works.

I keep forgetting about 'mdl', I defer to the spec itself as mdl is not always compliant, and I prefer compliance...  >.>

> As for the resolving to Chips, it should be doable just having a container with the chips and a textfield.

Yeah just an easy form that is to be repeated often, would be useful to have that as another new component (CardContainer?) perhaps someday.

**Pull request merged**